### DISCUSSION TITLE
How to get the top and bottom surface of a geo layer

### DISCUSSION THREAD

**Discussion by Jimmy-KL** [pyvista/pyvista#4042] [Q&A] [answered]
I have a geological layer as below (PolyData):
![image](https://user-images.githubusercontent.com/114913030/219957012-7d88e236-23fd-4a77-b24a-e352484ceba5.png)
In this layer, the side surfaces are always vertical. So is there a filter or a combination of filters to get the top surface, the bottom surface, and the side surfaces?

**Comment by ihsienlee** [comment]
You can use [clip_surface](https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSetFilters.clip_surface.html) to get surface.

**Reply by Jimmy-KL** [reply]
@ihsienlee I think `clip_surface` is not the best solution, because it's not easy to find the right surface to clip the model to get the top surface or bottom surface.

**Comment by banesullivan** [comment, marked as answer]
@Jimmy-KL, can you share this particular mesh?

I think your best bet is to cut the mesh along the sharp edges. I'm pretty sure an example/question like this has come up before but I'm struggling to dig anything up or recall how best to do this.

You could also compute the surface normals and then threshold based on an orientation range (getting only the upward-facing surface)

HTH

**Comment by Jimmy-KL** [comment]
[geo_layer.zip](https://github.com/pyvista/pyvista/files/10778818/geo_layer.zip)
@banesullivan I have attached this mesh above.
Yes, I thought about how to cut the sharp edges, for example, cutting the cells of which the angles between the horizontal plane is bigger than some degree. I found a filter called [PolyDataFilters.edge_mask](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.edge_mask.html). I tried it, but it seems it's not what I need. BTW, what is the exact definition of the argument - `surface angle`?
Using normals can be a solution I think. I will try, thanks!
